gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java

```diff
@@ -17,202 +17,156 @@
+      if (numRead == -1) {
+        response = -1;
+      } else {
+        if (numRead != 1) {
+          throw new IOException(
+              String.format(
+                  "Read %d bytes using single-byte buffer for path %s ending in position %d!",
+                  numRead, gcsPath, channel.position()));
         }
+        totalBytesRead++;
+        statistics.incrementBytesRead(1);
+        statistics.incrementReadOps(1);
+        response = (singleReadBuf[0] & 0xff);
       }
-      response = result;
     } catch (IOException e) {
       streamStatistics.readException();
       throw e;
     }
-    streamStatistics.bytesRead(1);
-    streamStatistics.readOperationCompleted(1, response);
+    streamStatistics.bytesRead(response == -1 ? 0 : 1);
+    streamStatistics.readOperationCompleted(/* requested= */ 1, response == -1 ? 0 : 1);
     return response;
   }
 
   @Override
-  public synchronized int read(byte[] buf, int offset, int length) throws IOException {
+  public synchronized int read(@Nonnull byte[] buf, int offset, int length) throws IOException {
     streamStatistics.readOperationStarted(getPos(), length);
     int response = 0;
     try {
-      int result;
-
-      synchronized (this) {
-        checkNotNull(buf, "buf must not be null");
-        if (offset < 0 || length < 0 || length > buf.length - offset) {
-          throw new IndexOutOfBoundsException();
-        }
-        int numRead = channel.read(ByteBuffer.wrap(buf, offset, length));
-        if (numRead > 0) {
-          // -1 means we actually read 0 bytes, but requested at least one byte.
-          totalBytesRead += numRead;
-          statistics.incrementBytesRead(numRead);
-          statistics.incrementReadOps(1);
-        }
-        result = numRead;
+      checkNotNull(buf, "buf must not be null");
+      if (offset < 0 || length < 0 || length > buf.length - offset) {
+        throw new IndexOutOfBoundsException();
       }
-      response = result;
+      int numRead = channel.read(ByteBuffer.wrap(buf, offset, length));
+      if (numRead > 0) {
+        // -1 means we actually read 0 bytes, but requested at least one byte.
+        totalBytesRead += numRead;
+        statistics.incrementBytesRead(numRead);
+        statistics.incrementReadOps(1);
+      }
+      response = numRead;
     } catch (IOException e) {
       streamStatistics.readException();
     }
-    streamStatistics.bytesRead(response);
-    streamStatistics.readOperationCompleted(length, response);
+    streamStatistics.bytesRead(max(response, 0));
+    streamStatistics.readOperationCompleted(length, max(response, 0));
     return response;
   }
 
   @Override
   public synchronized void seek(long pos) throws IOException {
+    logger.atFiner().log("seek(%d)", pos);
     long curPos = getPos();
     long diff = pos - curPos;
     if (diff > 0) {
       streamStatistics.seekForwards(diff);
     } else {
       streamStatistics.seekBackwards(diff);
     }
-    synchronized (this) {
```

Comment:
Any particular reason we are removing the synchronized block ( I am unaware of why it was there in the first place )

Reply:
This block is already in `synchronized` method, so this block is redundant.